### PR TITLE
Add a Heat aware DistributedSampler for torch usage.

### PR DESCRIPTION
**Pull request by Berkant03**
… on a process local basis.

## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [ ] unit tests: all split configurations tested
    - [ ] unit tests: multiple dtypes tested
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [ ] documentation updated where needed

## Description

Issue/s resolved: #1789 

## Changes proposed:

Add a Heat aware DistributedSampler for usage for PyTorch use cases  

## Type of change
- New feature (non-breaking change which adds functionality)

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by Berkant03** [comment]
When using the normal comm.Bcast the Bcast only works the first time and the seconds time not anymore.
```shell
❯ mpirun -np 2 python test.py
0 tensor([2, 4, 3, 0, 1], dtype=torch.int32)
1 tensor([2, 4, 3, 0, 1], dtype=torch.int32)
DNDarray([[10, 11, 12, 13, 14],
          [20, 21, 22, 23, 24],
          [15, 16, 17, 18, 19],
          [ 0,  1,  2,  3,  4],
          [ 5,  6,  7,  8,  9]], dtype=ht.int32, device=cpu:0, split=0)

1 tensor([455,   0,   0,   0,  32], dtype=torch.int32)
0 tensor([1, 4, 3, 2, 0], dtype=torch.int32)
```

**Comment by Berkant03** [comment]
The issue of the batches test is because of the way the iterator of the dataloader is created. Pytorch uses a wrapper Iterator for the Sampler and sets a random seed before the iterator of the sampler is invoked, therefore not using the provided seed. Because of this the samples do not match the shuffled reference array. Therefore i would just say to skip it.

**Comment by JuanPedroGHM** [comment]
**TODO**
- ```__repr__``` might have been a mistake, should return to previous behaviour :cry: 
- Investigate difference between existing ```ht.random.permute``` and ```_alltoall_shuffle```
- Move ```SeedEnvironemnt``` to base test class. (Creation and teardown methods, possibly)

**Review by JuanPedroGHM — APPROVED** [review]

File: heat/core/tests/test_suites/basic_test.py [hunk 1]
```diff
@@ -63,6 +63,7 @@ def setUpClass(cls) -> None:
             )
 
         cls.device, cls.other_device, cls.envar, cls.is_mps = ht_device, other_device, envar, is_mps
+        seed(42)
```
Comment (mrfh92) [inline heat/core/tests/test_suites/basic_test.py]: In my understanding, this line can be removed as with #1947 a seed has already been set in lines 21/22.

**Review by mrfh92 — DISMISSED** [review]
Looks fine to me. I only have the one comment below.
